Start on a Legacy submodule for backwards compat

tecosaur opened the pull request:
As we change more code to use the StyledStrings approach, it would be nice to try to maintain compatibility with old user customisations as much as possible.

This is currently an incomplete solution, but as a start we can add the capability to load 256-color values and respect (some of) the `JULIA_*_COLOR` environment variables.

StefanKarpinski commented on src/legacy.jl:
```diff
@@ -0,0 +1,126 @@
+    :light_cyan => :bright_cyan,
+    :light_white => :bright_white)
+
+legacy_color(color256::Int) = if 0 <= color256 <= 256 SimpleColor(ANSI_256_COLORS[color256]) end
```
```suggestion
legacy_color(color256::Int) = 0 <= color256 <= 256 ? SimpleColor(ANSI_256_COLORS[color256]) : nothing
```
This is more obvious to me; otherwise you have to know that an if statement without an else branch evaluates to `nothing` if the condition isn't true.

tecosaur replied:
I added an `else nothing` clause, and did the same with the `::String` method too.

StefanKarpinski reviewed (commented):
Is this code copied from somewhere or just emulating the old stuff with a new implementation? If it's copied code, then leaving it as identical as possible is good (for comparability); if it's new code emulating old behavior, it makes more sense to comment on style.

tecosaur commented:
This is new code to emulate prior behaviour, so I'm happy to take pointers on style.

StefanKarpinski commented:
Don't care for the ternary operator, huh?

StefanKarpinski commented on src/legacy.jl:
```diff
@@ -0,0 +1,135 @@
+A mapping from 256-color codes to 8-bit colours, represented with
+`SimpleColor`-compatible `UInt32`s.
+"""
+const ANSI_256_COLORS = Dict{Int, UInt32}(
```
Wouldn't this be better as a `UInt32` vector with 256 elements? When your dict keys are contiguous integers, what you actually want is an array.

tecosaur replied:
Ah yes, why didn't I think to do that initially? :laughing:

StefanKarpinski commented on src/legacy.jl:
```diff
@@ -0,0 +1,135 @@
+
+legacy_color(color::Symbol) = SimpleColor(get(RENAMED_COLORS, color, color))
+
+function legacy_color(color256::Int)
+    if 0 <= color256 <= 256
+        SimpleColor(ANSI_256_COLORS[color256])
+    else
+        nothing
+    end
+end
```
```suggestion
legacy_color(color256::Int) = get(ANSI_256_COLORS, color256, nothing)
```
The same API works for dicts and vectors, so this simplification is independent of the other suggestion.

tecosaur replied:
It still needs to be wrapped in `SimpleColor` should it be valid, so I've left this but changed the condition to `checkindex(Bool, ANSI_256_COLORS, color256+1)`.

StefanKarpinski commented on src/legacy.jl:
```diff
@@ -0,0 +1,135 @@
+                    "light_blue", "light_magenta", "light_cyan", "light_white")
+    if color in namedcolours
+        legacy_color(Symbol(namedcolours))
+    elseif 0 <= (color256 = something(tryparse(Int, ecolor), -1)) <= 255
```
Is `ecolor` a typo here?

tecosaur replied:
Yep, or more accurately a copy-paste artifact :sweat_smile:.

StefanKarpinski reviewed (changes requested):
A couple of suggested changes, one possible typo; maybe needs more tests so that typos would be caught?

tecosaur commented:
Thanks for giving this another/closer look Stefan. I appreciate you giving some of your time to this.

Just addressed all of your comments, since it seems like the general idea/design here is fine I'll write some tests for this before thinking of merging.

tecosaur commented:
Just added tests, and fixed two issues that became apparent over the course of doing so.

Since there have been no comments against this, other than code style improvements (which have been acted on), I'll be going ahead and merging this shortly.
